File: 54.spiral-matrix.py
#
# @lc app=leetcode id=54 lang=python3
#
# [54] Spiral Matrix
#

# @lc code=start
from collections import deque


class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        t, b = 0, len(matrix)
        l, r = 0, len(matrix[0])
        res = deque()
        while l < r and t < b:
            res.extend(matrix[t][i] for i in range(l, r))
            t += 1
            res.extend(matrix[i][r - 1] for i in range(t, b))
            r -= 1
            if l == r or t == b:
                break
            res.extend(matrix[b - 1][i] for i in reversed(range(l, r)))
            b -= 1
            res.extend(matrix[i][l] for i in reversed(range(t, b)))
            l += 1
        return res


# @lc code=end

File: RustSolutions/src/54.spiral-matrix.rs
/*
 * @lc app=leetcode id=54 lang=rust
 *
 * [54] Spiral Matrix
 */

// @lc code=start
impl Solution {
    pub fn spiral_order(matrix: Vec<Vec<i32>>) -> Vec<i32> {
        let (mut t, mut b) = (0, matrix.len());
        let (mut l, mut r) = (0, matrix[0].len());
        let mut res = vec![];
        
        while l < r && t < b {
            res.extend(matrix[t][l..r].iter());
            t += 1;
            res.extend((t..b).map(|i| matrix[i][r - 1]));
            r -= 1;
            if l == r || t == b {
                return res;
            }
            res.extend(matrix[b - 1][l..r].iter().rev());
            b -= 1;
            res.extend((t..b).map(|i| matrix[i][l]).rev());
            l += 1;
        }
        res
    }
}
// @lc code=end

